File: example/tracks.py
#!/usr/bin/env python
"""A program to show CD information."""
#
#  Copyright (C) 2006, 2008 Rocky Bernstein <rocky@gnu.org>
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.

import os, sys
libdir = os.path.join(os.path.dirname(__file__), '..')
if libdir[-1] != os.path.sep:
    libdir += os.path.sep
sys.path.insert(0, libdir)
import pycdio
import cdio

if sys.argv[1:]:
    try:
        d = cdio.Device(sys.argv[1])
    except IOError:
        print "Problem opening CD-ROM: %s" % sys.argv[1]
        sys.exit(1)
else:
    try:
        d = cdio.Device(driver_id=pycdio.DRIVER_UNKNOWN)
    except IOError:
        print "Problem finding a CD-ROM"
        sys.exit(1)

t = d.get_first_track()
if t is None: 
    print "Problem getting first track"
    sys.exit(2)

first_track = t.track
num_tracks  = d.get_num_tracks()
last_track  = first_track+num_tracks-1

try:
    last_session = d.get_last_session()
    print "CD-ROM %s has %d track(s) and %d session(s)." % \
          (d.get_device(), d.get_num_tracks(), last_session)
except cdio.DriverUnsupportedError:
    print "CD-ROM %s has %d track(s). " % (d.get_device(), d.get_num_tracks())

print "Track format is %s." %  d.get_disc_mode()

mcn = d.get_mcn()
if mcn:
    print "Media Catalog Number: %s" % mcn 
    
print "%3s: %-6s  %s" % ("#", "LSN", "Format")
i=first_track
while (i <= last_track):
    try: 
        t = d.get_track(i)
        print "%3d: %06lu  %-6s %s"  % (t.track, t.get_lsn(),
                                        t.get_msf(), t.get_format())
    except cdio.TrackError:
        pass
    i += 1

print "%3X: %06lu  leadout" \
      % (pycdio.CDROM_LEADOUT_TRACK, d.get_disc_last_lsn())
d.close()

File: example/tracks.c
/*
  $Id: tracks.c,v 1.1 2004/10/10 00:21:08 rocky Exp $

  Copyright (C) 2003, 2004 Rocky Bernstein <rocky@panix.com>
  
  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.
  
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

/* Simple program to list track numbers and logical sector numbers of
   a Compact Disc using libcdio. */
#ifdef HAVE_CONFIG_H
# include "config.h"
#endif
#include <stdio.h>
#include <sys/types.h>
#include <cdio/cdio.h>
int
main(int argc, const char *argv[])
{
  CdIo *p_cdio = cdio_open (NULL, DRIVER_UNKNOWN);
  track_t first_track_num;
  track_t num_tracks;
  int j, i=first_track_num;
  

  if (NULL == p_cdio) {
    printf("Couldn't find a driver.. leaving.\n");
    return 1;
  }
  
  num_tracks      = cdio_get_num_tracks(p_cdio);
  first_track_num = i = cdio_get_first_track_num(p_cdio);

  printf("CD-ROM Track List (%i - %i)\n", first_track_num, num_tracks);

  printf("  #:  LSN\n");
  
  for (j = 0; j < num_tracks; i++, j++) {
    lsn_t lsn = cdio_get_track_lsn(p_cdio, i);
    if (CDIO_INVALID_LSN != lsn)
	printf("%3d: %06lu\n", (int) i, (long unsigned int) lsn);
  }
  printf("%3X: %06lu  leadout\n", CDIO_CDROM_LEADOUT_TRACK, 
       (long unsigned int) cdio_get_track_lsn(p_cdio, 
					      CDIO_CDROM_LEADOUT_TRACK));
  cdio_destroy(p_cdio);
  return 0;
}
